################################################################################
SHORT DESCRIPTION: 
################################################################################
Use IO port 0xe9 in printf() to log directly to Xen console, like HVMLoader.

################################################################################
LONG DESCRIPTION: 
################################################################################
VGABIOS writes to IO port 0x500 by default in printf(), but this is not
available in our guest environment. Instead use 0xe9 handled by Xen to output
directly to Xen console like HVMLoader already does.

################################################################################
CHANGELOG 
################################################################################
Ported to VGABIOS: Eric Chanudet, chanudete@ainfosec.com, 26/03/2015

################################################################################
REMOVAL 
################################################################################
This patch can be removed on its own.

################################################################################
UPSTREAM PLAN 
################################################################################
This is OpenXT specific. It is already in VGABIOS sources distributed with Xen.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
None.

################################################################################
PATCHES
################################################################################
--- a/vgabios.c
+++ b/vgabios.c
@@ -4082,9 +4082,9 @@ void printf(s)
         for (i=0; i<format_width; i++) {
           nibble = (arg >> (4 * digit)) & 0x000f;
           if (nibble <= 9)
-            outb(0x0500, nibble + '0');
+            outb(0xe9, nibble + '0');
           else
-            outb(0x0500, (nibble - 10) + 'A');
+            outb(0xe9, (nibble - 10) + 'A');
           digit--;
           }
         in_format = 0;
@@ -4094,7 +4094,7 @@ void printf(s)
       //  }
       }
     else {
-      outb(0x0500, c);
+      outb(0xe9, c);
       }
     s ++;
     }
